### PR TITLE
Dissimilarity

### PR DESCRIPTION
**Pull request by petrelharp**
Here's a start at moving over the similarity/dissimilarity stuff from tsdate. WIP.

### CODE REVIEW

**Review by jeromekelleher — APPROVED** [review]
LGTM, happy to merge whenever

File: tscompare/methods.py [hunk 1]
```diff
@@ -0,0 +1,405 @@
...
+import numpy as np
+import scipy.sparse
+
+import tskit
+
+def node_spans(ts):
+    """
+    Returns the array of "node spans", i.e., the `j`th entry gives
+    the total span over which node `j` is in the tree (i.e., does
+    not have 'missing data' there).
```
Comment (petrelharp) [inline tscompare/methods.py:40]:
```suggestion
    the total span over which node `j` is in the tree sequence
    (i.e., does not have 'missing data' there).

```

File: tscompare/methods.py [hunk 2]
```diff
@@ -0,0 +1,405 @@
...
+        return clade in self._nodes
+
+
+def shared_node_spans(ts, other):
+    """
+    Calculate the spans over which pairs of nodes in two tree sequences are
+    ancestral to indentical sets of samples.
+
+    Returns a sparse matrix where rows correspond to nodes in `ts` and columns
+    correspond to nodes in `other`.
```
Comment (petrelharp) [inline tscompare/methods.py:174]:
```suggestion
    correspond to nodes in `other`, and whose value is the total amount of span
    over which the set of samples inheriting from the two nodes is identical.

    :return: A sparse matrix of class `scipy.sparse.csr_matrix`.
```

File: tscompare/methods.py [hunk 3]
```diff
@@ -0,0 +1,405 @@
...
+    ).tocsr()
+
+    return numer
+
+
+def match_node_ages(ts, other):
+    """
+    For each node in `ts`, return the age of a matched node from `other`.  Node
+    matching is accomplished by calculating the intervals over which pairs of
+    nodes (one from `ts`, one from `other`) subtend the same set of samples.
```
Comment (petrelharp) [inline tscompare/methods.py:254]:
```suggestion
    For each node in `ts`, return the age of a matched node from `other`.  Node
    matching is accomplished as described in :func:`.compare`.
    
```

File: tscompare/methods.py [hunk 4]
```diff
@@ -0,0 +1,405 @@
...
+
+    Returns three vectors of length `ts.num_nodes`: the age of the best
+    matching node in `other` (e.g.  with the longest shared span); the
+    proportion of the node span in `ts` that is covered by the best match; and
+    the id of the best match in `other`.
```
Comment (petrelharp) [inline tscompare/methods.py:259]:
```suggestion
    Returns a tuple of three vectors of length `ts.num_nodes`, in this order: 
    the age of the best matching node in `other`;
    the proportion of the node span in `ts` that is covered by the best match;
    and the node id of the best match in `other`.

```

File: tscompare/methods.py [hunk 5]
```diff
@@ -0,0 +1,405 @@
...
+
+    If either tree sequence contains unary nodes, then there may be multiple
+    matches with the same span for a single node. In this case, the returned
+    match is the node with the smallest integer id.
```
Comment (petrelharp) [inline tscompare/methods.py:263]:
```suggestion

:return: A tuple of arrays of length `ts.num_nodes` containing
    (time of matching node, proportion overlap, and node ID of match).
```

File: tscompare/methods.py [hunk 6]
```diff
@@ -0,0 +1,405 @@
...
+    """
+
+    shared_spans = shared_node_spans(ts, other)
```
Comment (petrelharp) [inline tscompare/methods.py:266]: TODO: refactor `compare` to use this (but doing tie-resolving using times as in that function).

File: tscompare/methods.py [hunk 7]
```diff
@@ -0,0 +1,405 @@
...
+    matched_time = other.nodes_time[best_match]
+
+    best_match[matched_span == 0] = tskit.NULL
+    matched_time[matched_span == 0] = np.nan
+
+    return matched_time, matched_span, best_match
+
+
+@dataclass
+class DissimilarityResult:
```
Comment (petrelharp) [inline tscompare/methods.py:280]:
```suggestion
class ARFResult:

```

File: tscompare/methods.py [hunk 8]
```diff
@@ -0,0 +1,405 @@
...
+    """
+    The result of a call to tscompare.dissimilarity(ts, other).
```
Comment (petrelharp) [inline tscompare/methods.py:282]:
```suggestion
    The result of a call to tscompare.compare(ts, other),
    returning metrics associated with the ARG Robinson-Foulds
    measures of similarity and dissimilarity.

```

File: tscompare/methods.py [hunk 9]
```diff
@@ -0,0 +1,405 @@
...
+    """
+    relative_dissimilarity: float
```
Comment (petrelharp) [inline tscompare/methods.py:284]:
```suggestion
    arf: float

```

File: tscompare/methods.py [hunk 10]
```diff
@@ -0,0 +1,405 @@
...
+    """
+    The proportion of the total span of ts that is *not* represented in other.
```
Comment (petrelharp) [inline tscompare/methods.py:286]:
```suggestion
    The ARG Robinson-Foulds relative dissimilarity:
    the proportion of the total span of `ts` that is *not* represented in `other`.

```

File: tscompare/methods.py [hunk 11]
```diff
@@ -0,0 +1,405 @@
...
+    This is: dissimilarity / total_span[0]
+    """
+    relative_similarity: float
```
Comment (petrelharp) [inline tscompare/methods.py:289]:
```suggestion
    tpr: float

```

File: tscompare/methods.py [hunk 12]
```diff
@@ -0,0 +1,405 @@
...
+    """
+    The proportion of the total span of other that is represented in ts.
```
Comment (petrelharp) [inline tscompare/methods.py:291]:
```suggestion
    The "true proportion represented":
    the proportion of the total span of `other` that is represented in `ts`.

```

File: tscompare/methods.py [hunk 13]
```diff
@@ -0,0 +1,405 @@
...
+    This is: (total_span[0] - dissimilarity) / total_span[1]
+    """
+    dissimilarity: float
+    """
+    The total span of ts that is not represented in other.
```
Comment (petrelharp) [inline tscompare/methods.py:296]:
```suggestion
    The total span of `ts` that is not represented in `other`.

```

File: tscompare/methods.py [hunk 14]
```diff
@@ -0,0 +1,405 @@
...
+    """
+    total_span: tuple
+    """
+    The total spans of the two tree sequences, in order (ts, other).
```
Comment (petrelharp) [inline tscompare/methods.py:300]:
```suggestion
    The total of all node spans of the two tree sequences, in order (`ts`, `other`).

```

File: tscompare/methods.py [hunk 15]
```diff
@@ -0,0 +1,405 @@
...
+    """
+    rmse: float
+    """
+    The root-mean-squared error between the transformed times of the nodes in
+    ts and the transformed times of their best-matching nodes in other, with
+    the average taken weighting by span
```
Comment (petrelharp) [inline tscompare/methods.py:306]:
```suggestion
    The root-mean-squared error between the transformed times of the nodes in
    `ts` and the transformed times of their best-matching nodes in `other`, with
    the average taken weighting by span in `ts`.

```

File: tscompare/methods.py [hunk 16]
```diff
@@ -0,0 +1,405 @@
...
+    """
+    transform: callable
+    """
+    The transformation function used to transform times for computing `rmse`.
+    """
+
+
+
+
+def dissimilarity(ts, other, transform=None):
```
Comment (petrelharp) [inline tscompare/methods.py:316]:
```suggestion
def compare(ts, other, transform=None):

```

File: tscompare/methods.py [hunk 17]
```diff
@@ -0,0 +1,405 @@
...
+    sample set does not match its' best match's descendant sample set.
+    2. The root mean squared difference
+    between the transformed times of the nodes in `ts`
+    and transformed times of their best matching nodes in `other`,
+    with the average weighted by the nodes' spans in `ts`.
+    3. The proportion of the span in `other` that is correctly
+    represented in `ts` (i.e., the total matching span divided
+    by the total span of `other`).
+
+    The transformation for times is by default log(1 + t).
```
Comment (petrelharp) [inline tscompare/methods.py:330]:
```suggestion
    `transform` is used to transform times before computing root-mean-squared error
    (see :class:`.ARFResult`); the default is `log(1 + t)`.

```

File: tscompare/methods.py [hunk 18]
```diff
@@ -0,0 +1,405 @@
...
+    this method returns three values:
+    1. The fraction of the total span of `ts` over which each nodes' descendant
+    sample set does not match its' best match's descendant sample set.
+    2. The root mean squared difference
+    between the transformed times of the nodes in `ts`
+    and transformed times of their best matching nodes in `other`,
+    with the average weighted by the nodes' spans in `ts`.
+    3. The proportion of the span in `other` that is correctly
+    represented in `ts` (i.e., the total matching span divided
+    by the total span of `other`).
```
Comment (petrelharp) [inline tscompare/methods.py:328]:
```suggestion
    For two tree sequences `ts` and `other`,
    this method returns an object of type :class:`.ARFResult`.
    The values reported summarize the degree to which nodes in `ts`
    "match" corresponding nodes in `other`.
    
    To match nodes,
    for each node in `ts`, the best matching node(s) from `other`
    has the longest matching span using :func:`.shared_node_spans`.
    If there are multiple matches with the same longest shared span
    for a single node, the best match is the match that is closest in time.
    
    Then, :class:`.ARFResult` contains:
    1. The total "matching span", which is the total span of all nodes
    in `ts` over which each node is ancestral to the same set of samples
    as its best match in `other`.
    2. (`arf`) The fraction of the total span of `ts` over which each nodes' 
    descendant sample set does not match its' best match's descendant 
    sample set (i.e., the total *un*-matched span divided by the total
    span of `ts`).
    3. (`tpr`) The proportion of the span in `other` that is correctly
    represented in `ts` (i.e., the total matching span divided
    by the total span of `other`).
    4. The root mean squared difference
    between the transformed times of the nodes in `ts`
    and transformed times of their best matching nodes in `other`,
    with the average weighted by the nodes' spans in `ts`.
    5. The total node spans of `ts` and `other`.


```

File: tscompare/methods.py [hunk 19]
```diff
@@ -0,0 +1,405 @@
...
+
+    The transformation for times is by default log(1 + t).
+
+    This is done as follows:
+
+    For each node in `ts`, the best matching node(s) from `other`
+    has the longest matching span using `shared_node_spans`.
+    If there are multiple matches with the same longest shared span
+    for a single node, the best match is the match that is closest in time.
```
Comment (petrelharp) [inline tscompare/methods.py:337]:
```suggestion
```

File: tscompare/methods.py [hunk 20]
```diff
@@ -0,0 +1,405 @@
...
+
+    :param ts: The focal tree sequence.
+    :param other: The tree sequence we check for inclusion in.
```
Comment (petrelharp) [inline tscompare/methods.py:340]:
```suggestion
    :param other: The tree sequence we compare to.

```

File: tscompare/methods.py [hunk 21]
```diff
@@ -0,0 +1,405 @@
...
+    :param transform: A callable that can take an array of times and
+        return another array of numbers.
+    :return: The three quantities above.
+    :rtype: DissimilarityResult
```
Comment (petrelharp) [inline tscompare/methods.py:344]:
```suggestion
    :rtype: ARFResult

```

File: tscompare/methods.py [hunk 22]
```diff
@@ -0,0 +1,405 @@
...
+    time_matrix = scipy.sparse.csr_matrix(
+        (time_difference, (row_ind[match], col_ind[match])),
+        shape=(ts.num_nodes, other.num_nodes),
+    )
+    time_discrepancies = np.asarray(
+        time_matrix[np.arange(len(best_match)), best_match].reshape(-1)
+    )
+    product = np.multiply((time_discrepancies**2), ts_node_spans)
+    rmse = np.sqrt(np.sum(product) / total_span_ts)
+    return DissimilarityResult(
```
Comment (petrelharp) [inline tscompare/methods.py:397]:
```suggestion
    return ARFResult(

```

File: tscompare/methods.py [hunk 23]
```diff
@@ -0,0 +1,405 @@
...
+            relative_dissimilarity = 1.0 - total_match_span / total_span_ts,
+            relative_similarity = total_match_span / total_span_other,
```
Comment (petrelharp) [inline tscompare/methods.py:399]:
```suggestion
            arf = 1.0 - total_match_span / total_span_ts,
            tpr = total_match_span / total_span_other,

```

**Review by petrelharp — COMMENTED** [review]
Collaborative.

**Comment by petrelharp** [comment]
This is passing tests; I will merge.

**Pull request merged**